File: buzzvil-sdk-v6-ios/buzzvil-sdk-ios-objc/BenefitHub/BenefitHubViewController.m
#import "BenefitHubViewController.h"
@import BuzzvilSDK;

@interface BenefitHubViewController ()

@property (nonatomic, strong) BuzzBenefitHub *benefitHub;
@property (nonatomic, strong) UIStackView *stackView;
@property (nonatomic, strong) UIButton *showBenefitHubButton;
@property (nonatomic, strong) UIButton *showLuckyBoxButton;
@property (nonatomic, strong) UIButton *showMissionPackButton;
@property (nonatomic, strong) UIButton *showHistoryButton;

@end

@implementation BenefitHubViewController

- (void)viewWillAppear:(BOOL)animated {
  [super viewWillAppear:animated];
  
  self.navigationController.navigationBarHidden = NO;
}

- (void)viewDidLoad {
  [super viewDidLoad];
  
  [self setupView];
  [self setupLayout];
}

- (void)setupView {
  self.view.backgroundColor = [UIColor systemBackgroundColor];
  
  self.showBenefitHubButton = [self createButtonWithTitle:@"Show BenefitHub" action:@selector(showBenefitHub)];
  self.showLuckyBoxButton = [self createButtonWithTitle:@"Show LuckyBox" action:@selector(showLuckyBox)];
  self.showMissionPackButton = [self createButtonWithTitle:@"Show MissionPack" action:@selector(showMissionPack)];
  self.showHistoryButton = [self createButtonWithTitle:@"Show History" action:@selector(showHistory)];
  
  self.stackView = [[UIStackView alloc] initWithArrangedSubviews:@[self.showBenefitHubButton, self.showLuckyBoxButton, self.showMissionPackButton, self.showHistoryButton]];
  self.stackView.axis = UILayoutConstraintAxisVertical;
  self.stackView.spacing = 8;
  self.stackView.distribution = UIStackViewDistributionFillEqually;
  
  [self.view addSubview:self.stackView];
  
  NSArray *buttonsToSetupAppearance = @[self.showBenefitHubButton, self.showLuckyBoxButton, self.showMissionPackButton, self.showHistoryButton];
  
  for (UIButton *button in buttonsToSetupAppearance) {
    [self setupAppearanceOfButton:button];
  }
}

- (UIButton *)createButtonWithTitle:(NSString *)title action:(SEL)action {
  UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
  [button setTitle:title forState:UIControlStateNormal];
  [button addTarget:self action:action forControlEvents:UIControlEventTouchUpInside];
  return button;
}

- (void)setupAppearanceOfButton:(UIButton *)button {
  button.backgroundColor = [UIColor systemBlueColor];
  button.layer.cornerRadius = 8;
  [button setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
  button.titleLabel.font = [UIFont systemFontOfSize:16];
  button.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [button.heightAnchor constraintEqualToConstant:32]
  ]];
}

- (void)setupLayout {
  self.stackView.translatesAutoresizingMaskIntoConstraints = NO;
  [NSLayoutConstraint activateConstraints:@[
    [self.stackView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
    [self.stackView.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor],
    [self.stackView.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor],
    [self.stackView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor],
  ]];
}

- (void)showBenefitHub {
  self.benefitHub = [[BuzzBenefitHub alloc] init];
  [self.benefitHub showOn:self];
}

- (void)showLuckyBox {
  self.benefitHub = [[BuzzBenefitHub alloc] init];
  BuzzBenefitHubConfig * benefithubConfig = [BuzzBenefitHubConfig configWith:^(BuzzBenefitHubConfigBuilder * _Nonnull builder) {
    builder.queryParams = [BuzzBenefitHubPage.luckyBox toRedirectQueryParams];
  }];
  
  [self.benefitHub setConfig:benefithubConfig];
  [self.benefitHub showOn:self];
}

- (void)showMissionPack {
  self.benefitHub = [[BuzzBenefitHub alloc] init];
  BuzzBenefitHubConfig * benefithubConfig = [BuzzBenefitHubConfig configWith:^(BuzzBenefitHubConfigBuilder * _Nonnull builder) {
    builder.queryParams = [BuzzBenefitHubPage.missionPack toRedirectQueryParams];
  }];
  
  [self.benefitHub setConfig:benefithubConfig];
  [self.benefitHub showOn:self];
}

- (void)showHistory {
  self.benefitHub = [[BuzzBenefitHub alloc] init];
  BuzzBenefitHubConfig * benefithubConfig = [BuzzBenefitHubConfig configWith:^(BuzzBenefitHubConfigBuilder * _Nonnull builder) {
    builder.queryParams = [BuzzBenefitHubPage.history toRedirectQueryParams];
  }];
  
  [self.benefitHub setConfig:benefithubConfig];
  [self.benefitHub showOn:self];
}

@end

File: buzzvil-sdk-v6-ios/buzzvil-sdk-ios-swift/BenefitHub/BenefitHubViewController.swift
import UIKit
import BuzzvilSDK

class BenefitHubViewController: UIViewController {
  private lazy var stackView: UIStackView = {
    let stackView = UIStackView(arrangedSubviews: [showBenefitHubButton, showLuckyBoxButton, showMissionPackButton, showHistoryButton])
    stackView.axis = .vertical
    stackView.spacing = 8
    stackView.distribution = .fillEqually
    
    return stackView
  }()
  
  private lazy var showBenefitHubButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Show BenefitHub", for: .normal)
    button.addTarget(self, action: #selector(showBenefitHub), for: .touchUpInside)
    
    return button
  }()
  
  private lazy var showLuckyBoxButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Show LuckyBox", for: .normal)
    button.addTarget(self, action: #selector(showLuckyBox), for: .touchUpInside)

    return button
  }()
  
  private lazy var showMissionPackButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Show MissionPack", for: .normal)
    button.addTarget(self, action: #selector(showMissionPack), for: .touchUpInside)

    return button
  }()
  
  private lazy var showHistoryButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Show History", for: .normal)
    button.addTarget(self, action: #selector(showHistory), for: .touchUpInside)

    return button
  }()

  private lazy var benefitHub = BuzzBenefitHub()
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    setupView()
    setupLayout()
  }
  
  private func setupView() {
    view.backgroundColor = .systemBackground
    
    view.addSubview(stackView)
    
    let buttonsToSetupAppearance = [
      showBenefitHubButton,
      showLuckyBoxButton,
      showMissionPackButton,
      showHistoryButton,
    ]
    
    buttonsToSetupAppearance.forEach { button in
      setupAppearanceOfButton(button: button)
    }
  }
  
  private func setupAppearanceOfButton(button: UIButton) {
    button.backgroundColor = .systemBlue
    button.layer.cornerRadius = 8
    button.setTitleColor(.white, for: .normal)
    button.titleLabel?.font = UIFont.systemFont(ofSize: 16)
    button.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      button.heightAnchor.constraint(equalToConstant: 32)
    ])
  }
  
  private func setupLayout() {
    stackView.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      stackView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
      stackView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
      stackView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
      stackView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor),
    ])
  }
  
  // 베네핏허브 표시하기
  @objc
  private func showBenefitHub() {
    benefitHub = BuzzBenefitHub()
    
    benefitHub.show(on: self)
  }
  
  // 럭키박스 표시하기
  @objc
  private func showLuckyBox() {
    benefitHub = BuzzBenefitHub()
    let benefitHubConfig = BuzzBenefitHubConfig.Builder()
      .setQueryParams(BuzzBenefitHubPage.luckyBox.toRedirectQueryParams())
      .build()
    benefitHub.setConfig(benefitHubConfig)
    benefitHub.show(on: self)
  }
  
  // 미션팩 표시하기
  @objc
  private func showMissionPack() {
    benefitHub = BuzzBenefitHub()
    let benefitHubConfig = BuzzBenefitHubConfig.Builder()
      .setQueryParams(BuzzBenefitHubPage.missionPack.toRedirectQueryParams())
      .build()
    benefitHub.setConfig(benefitHubConfig)

    benefitHub.show(on: self)
  }
  
  // 적립내역 표시하기
  @objc
  private func showHistory() {
    benefitHub = BuzzBenefitHub()
    let benefitHubConfig = BuzzBenefitHubConfig.Builder()
      .setQueryParams(BuzzBenefitHubPage.history.toRedirectQueryParams())
      .build()
    benefitHub.setConfig(benefitHubConfig)
    
    benefitHub.show(on: self)
  }
}
